Client tests

MarhiievHE opened the pull request:
Update client test

tshemsedinov commented on application/api/files.1/upload.js:
```diff
@@ -1,5 +1,13 @@
 async ({ streamId, name }) => {
   const filePath = `./application/resources/${name}`;
+  let exist = false;
+  try {
+    await node.fsp.access(filePath);
+    exist = true;
+  } catch (err) {
+    console.error(err);
+  }
```
```suggestion
  let exist = false;
  const toBool = [() => true, () => false];
  const exists = await fs.promises.access(filePath).then(...toBool);
```

tshemsedinov commented on application/api/files.1/upload.js:
```diff
@@ -1,5 +1,13 @@
+  } catch (err) {
+    console.error(err);
+  }
+  if (exist) return { result: `File ${name} already exist on server` };
```
```suggestion
  if (exists) return { result: `File ${name} already exist on server` };
```

tshemsedinov reviewed (changes requested):
Is it subject to close https://github.com/metarhia/Example/pull/204 in favour of this PR?

MarhiievHE commented:
@tshemsedinov Perhaps partially. Maybe it will be better if @leonpolak looks at whether everything is present, what he expected. I did not look at PR, but just corrected what interfered with the work of the test.
PS: I gave access to changes to my Fork.

tshemsedinov commented on application/lib/task1/start.js:
```diff
@@ -1,13 +1,13 @@
-      });
-      console.log('Add task', res);
-    }, 1000);
+    // setTimeout(async () => {
+    //   const res = await application.scheduler.add({
+    //     name: 'name',
+    //     every: 'Sep 10th 10s',
+    //     args: { i: 2 },
+    //     run: 'lib.task1.f1',
+    //   });
+    //   console.log('Add task', res);
+    // }, 1000);
```
```suggestion
    setTimeout(async () => {
      const res = await application.scheduler.add({
        name: 'name',
        every: 'Sep 10th 10s',
        args: { i: 2 },
        run: 'lib.task1.f1',
      });
      console.log('Add task', res);
    }, 1000);
```

tshemsedinov commented on application/lib/task1/start.js:
```diff
@@ -1,13 +1,13 @@
 async () => {
   if (application.worker.id === 'W1') {
```
```suggestion
  if (application) return; // comment to enable
  if (application.worker.id === 'W1') {
```

tshemsedinov replied:
I'd prefer this style to enable/disable such examples

tshemsedinov replied:
You can add this in other `start` handlers

tshemsedinov replied:
I think disable such features are also not a subject of this PR and added just to make all project runable

tshemsedinov commented on test/client.js:
```diff
@@ -1,236 +1,323 @@
+) => {
+  const tests = {
+    'system/introspect': async (test) => {
+      const introspect = await wsClient.scaffold('system')('introspect')([
+        'auth',
+        'console',
+        'example',
+        'files',
+        'test',
+      ]);
```
```suggestion
      const units = ['auth', 'console', 'example', 'files', 'test'];
      const introspect = await wsClient.scaffold('system')('introspect')(units);
```

tshemsedinov commented on test/client.js:
```diff
@@ -1,236 +1,323 @@
+      try {
+        const resources = await wsApi.example.resources();
+        test.strictEqual(resources?.total, null);
+        // console.log({resources});
```
Final version before landing should not contain commented console.log's

MarhiievHE replied:
ok, do I need to remove all commented code too?

tshemsedinov replied:
Just before landing, or you can add comment `// not for landing` if you added something not related to this PR or commented something just to make this code works but if it should be subject to separate PR

tshemsedinov commented:
Looks like redis can't be accessed in docker: `unhandledRejection: Error: connect ECONNREFUSED 127.0.0.1:6379`
@MarhiievHE

MarhiievHE commented:
> Looks like redis can't be accessed in docker: `unhandledRejection: Error: connect ECONNREFUSED 127.0.0.1:6379` @MarhiievHE

You are right, I have forgotten to add optional support running API under a container with Redis. Ok, I will add a config for Redis.

MarhiievHE commented:
@tshemsedinov Do you have any other suggestions to improve or add some more tests?

tshemsedinov commented:
> @tshemsedinov Do you have any other suggestions to improve or add some more tests?

Tests are still failing https://github.com/metarhia/Example/pull/225/checks

MarhiievHE commented:
> > @tshemsedinov Do you have any other suggestions to improve or add some more tests?
> 
> Tests are still failing https://github.com/metarhia/Example/pull/225/checks
It`s because the test.yml haven`t any instance of Redis in the test environment.

tshemsedinov commented:
Please rebase on master

leonpolak reviewed (commented):
Sorry for such a massive delay in the review. Everything works and all required tests pass (Huge respect to Heorhii and Timur).
Just 3 questions:
1) Where is uploaded content stored?
2) Is 'example/subscribe' should work like this? Shouldn't it listen indefinitely for server-side events, until some 'unsubscribe' call?
3) I get 'warning: MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added to [MetaReadable]'. I obviously can use emitter.setMaxListeners(11), but it does not feel right.

MarhiievHE commented:
> Sorry for such a massive delay in the review. Everything works and all required tests pass (Huge respect to Heorhii and Timur). Just 3 questions:
> 
> 1. Where is uploaded content stored?
> 2. Is 'example/subscribe' should work like this? Shouldn't it listen indefinitely for server-side events, until some 'unsubscribe' call?
> 3. I get 'warning: MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added to [MetaReadable]'. I obviously can use emitter.setMaxListeners(11), but it does not feel right.

You're welcome.
1. There are two examples of file upload. The first through the Metacom stream saves to the path /application/resources/. The second through serialization, saves to the path application/tmp/.
2. I agree will be well expanding the algorithm.
3. Perhaps this version of the code refers to an old Metacom not yet been corrected. It will be necessary to apply changes from the master and see if the problem remains.

PS: I was a little busy in August, I was preparing for a language exam, and now I want to return to rewriting the Metatest library on the node test runner. But on Thursday, I plan to be on the community call and it will be possible to discuss what is not enough to close this PR.

MarhiievHE commented:
@leonpolak @tshemsedinov 
I rewrote the test to use a node test runner and fixed a subscription check. And I also corrected the GitHub Actions script. Now, all tests are completed correctly.

MarhiievHE commented:
@leonpolak Did you talk about this error before? I caught it now too.
<img width="1094" alt="image" src="https://github.com/metarhia/Example/assets/12165613/cb2c0a68-45ae-4a46-b1ba-20c1f99e67f8">
```
19:59:09  W2   error   warning: MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 Symbol() listeners added to [MetaReadable]. Use emitter.setMaxListeners() to increase limit
  _addListener (node:events:587:17)
  MetaReadable.addListener (node:events:605:10)
  MetaReadable.once (node:events:649:8)
  eventTargetAgnosticAddListener (node:events:1004:15)
  node:events:966:5
  new Promise (<anonymous>)
  Function.once (node:events:949:10)
  MetaReadable.push (/node_modules/metacom/lib/streams.js:45:26)
  Server.binary (/node_modules/metacom/lib/server.js:300:18)
```
